src/ui/Toolbar.tsx

```diff
@@ -10,166 +9,117 @@ type Props = {
   primitives?: PrimitiveGroup;
   languageId?: string; // used to find the .blocks-language-{languageId} CSS class
   blockMode?: boolean;
+  toolbarRef: React.RefObject<HTMLInputElement>;
 };
 
-type State = {
-  search: string;
-  selectedPrimitive: null | Primitive;
-};
-
-export default class Toolbar extends Component<Props, State> {
-  constructor(props: Props) {
-    super(props);
-    this.handleFocusPrimitive = this.handleFocusPrimitive.bind(this);
-    this.state = {
-      search: "",
-      selectedPrimitive: null,
-    };
-  }
-
-  static defaultProps: Props = {
-    primitives: null,
-    blockMode: false,
-  };
-
-  primitiveSearch: HTMLElement;
-
-  changeSearch: React.ChangeEventHandler<HTMLInputElement> = (event) => {
-    this.setState({ search: event.target.value });
-  };
-
-  clearSearch = () => {
-    this.setState({ search: "" });
-  };
-
-  next() {
-    let primitives = this.getPrimitives();
-    if (primitives.length == 0) return; // Nothing to select.
-    let i = this.getSelectedPrimitiveIndex(primitives);
-    // If nothing is selected, select the first primitive, otherwise select the next.
-    if (i === primitives.length - 1) i -= 1;
-    if (i === null) i = -1;
-    this.selectPrimitive(primitives[i + 1]);
-  }
-
-  prev() {
-    let primitives = this.getPrimitives();
-    if (primitives.length == 0) return; // Nothing to select.
-    let i = this.getSelectedPrimitiveIndex(primitives);
-    // If the first primitive is selected, keep it, otherwise select the previous.
-    if (i === 0) i = 1;
-    this.selectPrimitive(i === null ? null : primitives[i - 1]);
-  }
-
-  handleFocusPrimitive(selectedPrimitive: State["selectedPrimitive"]) {
-    this.setState({ selectedPrimitive: selectedPrimitive });
-  }
-
-  selectPrimitive(selectedPrimitive: null | Primitive) {
-    if (selectedPrimitive?.element) {
-      selectedPrimitive.element.focus(); // will trigger handleFocusPrimitive
+var Toolbar = (props: Props) => {
+  const { primitives, languageId, blockMode, toolbarRef } = props;
+
+  const [search, setSearch] = useState("");
+  const [selectedPrimitive, setSelectedPrimitive] = useState(null);
+  const clearSearch = () => setSearch("");
+  const changeSearch: React.ChangeEventHandler<HTMLInputElement> = (event) =>
+    setSearch(event.target.value);
+
+  // Get a flat array of all primitives matching 'search'
+  const getPrimitives = () =>
+    (primitives?.filter(search).primitives || []) as Primitive[];
+
+  // Set selectedPrimitive state, depending on whether we go up or down
+  const move = (dir: string) => {
```

Comment:
If you want to be extra clear, you could specify the  type of `dir` as `"Up"|"Down"`

Reply:
Ooooh, enumerator types?? Done and done!